test/helper/common.bash

```diff
@@ -117,131 +232,207 @@ function repeat_in_container_until_success_or_timeout() {
+function _wait_until_account_maildir_exists() {
+  local CONTAINER_NAME=${1:?Container name must be provided}
+  local MAIL_ACCOUNT=${2:?Mail account must be provided}
 
   local LOCAL_PART="${MAIL_ACCOUNT%@*}"
   local DOMAIN_PART="${MAIL_ACCOUNT#*@}"
   local MAIL_ACCOUNT_STORAGE_DIR="/var/mail/${DOMAIN_PART}/${LOCAL_PART}"
 
-  repeat_in_container_until_success_or_timeout 60 "${CONTAINER_NAME}" bash -c "[[ -d ${MAIL_ACCOUNT_STORAGE_DIR} ]]"
+  _repeat_in_container_until_success_or_timeout 60 "${CONTAINER_NAME}" \
+    /bin/bash -c "[[ -d ${MAIL_ACCOUNT_STORAGE_DIR} ]]"
+}
+
+# Wait until the mail queue is empty inside a container (${1}).
+#
+# @param ${1} = container name
+function _wait_for_empty_mail_queue_in_container() {
+  local CONTAINER_NAME=${1:?Container name must be provided}
+  local TIMEOUT=${TEST_TIMEOUT_IN_SECONDS}
+
+  # shellcheck disable=SC2016
+  _repeat_in_container_until_success_or_timeout \
+    "${TIMEOUT}" \
+    "${CONTAINER_NAME}" \
+    /bin/bash -c '[[ $(mailq) == *"Mail queue is empty"* ]]'
 }
 
-function add_mail_account_then_wait_until_ready() {
-  local CONTAINER_NAME=$1
-  local MAIL_ACCOUNT=$2
+
+# ? << Functions to wait until a condition is met
+# ! -------------------------------------------------------------------
+# ? >> Miscellaneous helper functions
+
+# Adds a mail account and waits for the associated files to be created.
+#
+# @param ${1} = container name
+# @param ${2} = account name
+# @param ${3} = password [OPTIONAL]
+function _add_mail_account_then_wait_until_ready() {
+  local CONTAINER_NAME=${1:?Container name must be provided}
+  local MAIL_ACCOUNT=${2:?Mail account must be provided}
   # Password is optional (omit when the password is not needed during the test)
   local MAIL_PASS="${3:-password_not_relevant_to_test}"
 
-  run docker exec "${CONTAINER_NAME}" setup email add "${MAIL_ACCOUNT}" "${MAIL_PASS}"
+  _run_in_container setup email add "${MAIL_ACCOUNT}" "${MAIL_PASS}"
   assert_success
 
-  wait_until_account_maildir_exists "${CONTAINER_NAME}" "${MAIL_ACCOUNT}"
+  _wait_until_account_maildir_exists "${CONTAINER_NAME}" "${MAIL_ACCOUNT}"
 }
 
-function wait_for_empty_mail_queue_in_container() {
-  local CONTAINER_NAME="${1}"
-  local TIMEOUT=${TEST_TIMEOUT_IN_SECONDS}
+# Assert that the number of lines output by a previous command matches the given
+# amount (${1}). `lines` is a special BATS variable updated via `run`.
+#
+# @param ${1} = number of lines that the output should have
+function _should_output_number_of_lines() {
+  assert_equal "${#lines[@]}" "${1}"
+}
 
-  # shellcheck disable=SC2016
-  repeat_in_container_until_success_or_timeout "${TIMEOUT}" "${CONTAINER_NAME}" bash -c '[[ $(mailq) == *"Mail queue is empty"* ]]'
+# Reloads the postfix service.
+#
+# @param ${1} = container name [OPTIONAL]
+function _reload_postfix() {
+  local CONTAINER_NAME=${1:-${CONTAINER_NAME}}
+
+  # Reloading Postfix config after modifying it in <2 sec will cause Postfix to delay, workaround that:
```

Comment:
Since this PR is quite noisy in diff (_not fun with tracking info on changes with `git blame`_), it might be helpful to reference the PR that details the workaround reasoning?

```suggestion
  # Reloading Postfix config after modifying it within 2 seconds will cause Postfix to delay reading `main.cf`:
  # WORKAROUND: https://github.com/docker-mailserver/docker-mailserver/pull/2998
```